From b4e10b6bc36bbd8bfb33bc6b74e8d3b1d3d75a9e Mon Sep 17 00:00:00 2001
From: Matthias Clasen <mclasen@redhat.com>
Date: Fri, 25 Nov 2022 07:59:35 -0500
Subject: [PATCH] text: Prevent unexpected notify::direction emissions

Calling gdk_device_get_direction can trigger a
notify::direction emissions, since the X11 backend
determines the direction on demand. Prevent that by
forcing the direction to be determined ahead of time.

Fixes: #5311
---
 gtk/gtktext.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/gtk/gtktext.c b/gtk/gtktext.c
index 89dda9782f..a8c21e2fd1 100644
--- a/gtk/gtktext.c
+++ b/gtk/gtktext.c
@@ -3237,8 +3237,12 @@ gtk_text_focus_changed (GtkEventControllerFocus *controller,
   if (gtk_event_controller_focus_is_focus (controller))
     {
       if (keyboard)
-        g_signal_connect (keyboard, "notify::direction",
-                          G_CALLBACK (direction_changed), self);
+        {
+          /* Work around unexpected notify::direction emissions */
+          gdk_device_get_direction (keyboard);
+          g_signal_connect (keyboard, "notify::direction",
+                            G_CALLBACK (direction_changed), self);
+        }
 
       gtk_text_im_set_focus_in (self);
       gtk_text_reset_blink_time (self);
-- 
GitLab

